README.md
# CashMaster

A currency management script for the D&D 5e OGL and 5e Shaped sheets on Roll20.net.

Please use `!cm` for inline help and examples.

## Setup

Make sure you use the correct sheet setting (`OGL`, `5E-Shaped`, or `other`).

## Usage

First, select one or several party members. 

### Base commands

- `!cm` or `!cm -help` or `!cm -h` will show this help overview
- `!cm -overview` or `!cm -o` to get an **overview** over the party's cash
	- `!cm -overview --usd` will also give you an overview and a rough conversion to USD (default value: 1 gp equals roughly 110 USD). 

### Payment commands

- `!cm -add [amount][currency]` or `!cm -a [amount][currency]` to **add** an equal amount of money to each selected party member,
- `!cm -loot [amount][currency]` or `!cm -l [amount][currency]` to **split up** a certain amount of coins between the party members, like a found treasure. Note that in this case, no conversion between the different coin types is made - if a party of 5 shares 4 pp, then 4 party members receive one pp each, and the last member won't get anything.
- `!cm -pay [amount][currency]` or `!cm -p [amount][currency]` to let each selected party member **pay** a certain amount. The script will even try to take higher and lower coin types to get the full amount. E.g. to pay 1gp when the character has no gold, the script will use 1pp (and return 9gp), or it will take 2ep, 10sp or 100cp - or any other valid combination of coins - to pay the desired amount.

### Conversion/Cleanup commands

- `!cm -share` or `!cm -s` to **convert and share** the money equally between party members, converting the amount into the best combination of gold, silver and copper (this should be used in smaller stores),
- `!cm -best-share` or `!cm -bs` to **convert and share** the money equally between party members, converting the amount into the best combination of platinum, gold, electrum, silver and copper (this should only be used in larger stores that have a fair amount of cash),


**Note:** You can use several coin values at once, e.g. `!cm -loot 50gp 150sp 2000cp` or `!cm -pay 2sp 5cp`.


### Examples

1. `!cm -overview` will show a cash overview.
2. `!cm -add 50gp` will add 50 gp to every selected character.
3. `!cm -loot 50gp` will (more or less evenly) distribute 50 gp among the party members.
4. `!cm -pay 10gp` will subtract 10gp from each selected character. It will try to exchange the other coin types (e.g. it will use 1pp if the player doesn't have 10gp).
5. `!cm -share` will collect all the money and share it evenly on the members, using gp, sp and cp only (pp and ep will be converted). Can also be used for one character to 'exchange' money.
6. `!cm -convert` - same as `!cm -share`, but will also use platinum and electrum.

## Credits

With thanks to [Kryx](https://app.roll20.net/users/277007/kryx)/[mlenser](https://github.com/mlenser) for his contributions.

CHANGELOG.md
# Release notes
## v0.1
**First alpha release**  
2018-01-09  
*Changes from `project start` to `v0.1`:*

* Added first routine to balance all accounts
* Adding roll template for nicer output
* Adding function to add/substract cash

## v0.2
**Second alpha**  
2018-01-09  
*Changes from `v0.1` to `v0.2`:*

* Improved output of cmadd

## v0.3
**Third alpha, first release**  
2018-01-10  
*Changes from `v0.2` to `v0.3`:*

* Started w readme
* Added help feature
* Small bugfix for cmshare
* Added another share option, updated readme
* Updated README
* Implemented some of Aaron's code
* Minor fix for help command
* Minor bug fix for negative amounts
* Improved USD conversion

## v0.4
**Second release**  
2018-01-11  
*Changes from `v0.3` to `v0.4`:*

* Minor bugfix for rest copper
* Added hoard function
* Bugfixes in the hoard routine
* Updated README

## v0.4.1
**Bugfixes, third release**  
2018-01-11  
*Changes from `v0.4` to `v0.4.1`:*

* Bugfix for negative treasure
* Merged cmshare and cmconvert routines
* Changed to attr.setWithWorker to trigger weight recalc

## v0.5
**Feature update**  
2018-01-20  
*Changes from `v0.4.1` to `v0.5`:*

* Added cmpay routine (+)
* Readme updated
* Adjusted conversion rate to 110 USD
* Added routine to check if characters were selected
* Small bugfix for cm_changemoney
* Stored the roll template in a variable
* Updating json
* Fixed the output for 5E-Shaped
* Adjusted script.json

## v0.6.0
**Maintenance update**  
2018-02-13  
*Changes from `v0.5` to `v0.6.0`:*

* Removed obsolete test file
* Change command syntax
* Update readme with new syntax
* Gulp init
* Started with gulp routine
* Adjust vnumber in json files
* Update gulp file
* Minor fix to json file
* GULP: Add jsonlint to gulp
* GULP: Copy content of publish into main repo
* Move output after transactions
* Fix the recognition of subcommands
* Improve README text
* initial cleanup
* consistent indentation and other cleanup (much more needed)
* parseFloat, not parseInt
* Add credits to mlsenser to README master
* Update files in publish
* remove node_modules from git
* add ignore files
* lint
* initial es6
* build command
* bump versions
* babelrc
* set up npm version bumping
* babel-preset-env
* yarn.lock
* babel and bump
* Fixing gulpfile
* Added support for general roll template
* Add json check to gulp lint
* Update useroptions in script.json
* Add gulp lintfix
* Cleaning up output
* Add weight calculation to output
* Bugfix in calculation routine
* Colorize output
* Add a check if attribute exists before changing value
* Missing 'else' added
* Fixed some lint errors
* Bugfix to check if token is selected
* Bumping prerelase vnumber
* Syntax proposal added
* Updating command syntax according to proposal
* Updating README to reflect syntax changes


gulpfile.js
/*
yarn
*/

const gulp = require('gulp');
const babel = require('gulp-babel');
const eslint = require('gulp-eslint');
const replace = require('gulp-string-replace');
const gfi = require('gulp-file-insert');
const markdown = require('gulp-markdown');
const del = require('del');
const bump = require('gulp-bump');
const jsonlint = require('gulp-jsonlint');
const p = require('./package.json');

const version = p.version.replace('v', '');

// ### Version number bump routines
//
// Basic usage:
// Will patch the version

const bumpVersion = (type) => {
  gulp.src('./package.json')
    .pipe(bump({ type }))
    .pipe(gulp.dest('./'));
};
gulp.task('bump-major', () => bumpVersion('major'));
gulp.task('bump-minor', () => bumpVersion('minor'));
gulp.task('bump-patch', () => bumpVersion('patch'));
gulp.task('bump-pre', () => bumpVersion('prepatch'));
gulp.task('bump-prerelease', () => bumpVersion('prerelease'));

gulp.task('lint', () => {
  gulp.src(['./*.js'])
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(eslint.failAfterError());
  gulp.src('./prep/script.json')
    .pipe(jsonlint())
    .pipe(jsonlint.reporter());
});

gulp.task('lintfix', () => {
  gulp.src(['./*.js'])
    .pipe(eslint({ fix: true }))
    .pipe(eslint.format())
    .pipe(gulp.dest('./'));
});


// prepare the three readme versions
// - a html version for inline help
// - a md with converted linebreaks for the json
// - and the clean md for the publish folder

gulp.task('pub1', () => {
  gulp.src('./README.md')
    .pipe(markdown())
    .pipe(replace(new RegExp('\n', 'g'), '')) // eslint-disable-line no-control-regex
    .pipe(replace(new RegExp('"', 'g'), '\''))
    .pipe(gulp.dest('./tmp/script'));

  gulp.src('./README.md')
    .pipe(replace(new RegExp('\n', 'g'), '\\r')) // eslint-disable-line no-control-regex
    .pipe(replace(new RegExp('"', 'g'), '\''))
    .pipe(gulp.dest('./tmp/package'));

  gulp.src('./README.md')
    .pipe(gulp.dest('./publish'));

  gulp.src('./prep/script.json')
    .pipe(jsonlint())
    .pipe(jsonlint.reporter());
});

// Then, replace vn and readme content in Cashmaster.js
// and script.json and write everything to the publish folder
// The package.json still needs to be updated manually a bit (older versions)

gulp.task('pub2', () => {
  gulp.src('./Cashmaster.js')
    .pipe(replace(new RegExp('%%version%%', 'g'), version))
    .pipe(gfi({
      '%%README%%': 'tmp/script/README.md',
    }))
    .pipe(babel({
      presets: ['env'],
    }))
    .pipe(gulp.dest('./publish'))
    .pipe(gulp.dest(`./publish/${version}`));

  gulp.src('./prep/script.json')
    .pipe(replace(new RegExp('%%version%%', 'g'), version))
    .pipe(gfi({
      '%%README%%': 'tmp/package/README.md',
    }))
    .pipe(gulp.dest('./publish'));

  gulp.src('./publish/script.json')
    .pipe(jsonlint())
    .pipe(jsonlint.reporter());
});

gulp.task('pub3', () => {
  gulp.src('./publish/script.json')
    .pipe(jsonlint())
    .pipe(jsonlint.reporter());

  gulp.src('./publish/*')
    .pipe(gulp.dest('../roll20-api-scripts/CashMaster/'));
  return del.sync('./tmp');
});

gulp.task('cleanup', () => {
  del.sync('./tmp');
});


gulp.task('cleanup2', () => {
  del.sync('./tmp');
  del.sync('./publish');
});


CashMaster.js
/* global on log playerIsGM findObjs getObj getAttrByName sendChat globalconfig */

/*
CASHMASTER %%version%%

A currency management script for the D&D 5e OGL sheets on roll20.net.
Please use `!cm` for inline help and examples.

arthurbauer@me.com
*/

const cashsplit = (c, m, x) => {
  //! cashsplit
  let ct = 0;
  let cr = 0;
  if (c !== null) {
    ct = Math.floor(c / m);
    cr = c % m;
    if (cr >= x || (c < 0 && cr < 0 && -cr < x)) {
      ct += 1;
    }
  }
  return ct;
};

const getattr = (cid, att) => {
  //! getattr
  const attr = findObjs({
    type: 'attribute',
    characterid: cid,
    name: att,
  })[0];
  if (attr) {
    return attr.get('current');
  }
  return '';
};

const setattr = (cid, att, val) => {
  //! setattr
  const attr = findObjs({
    type: 'attribute',
    characterid: cid,
    name: att,
  })[0];
  if (typeof attr === 'undefined' || attr == null) {
    const attr = createObj('attribute', { name: att, characterid: cid, current: parseFloat(val) }); // eslint-disable-line no-unused-vars, no-undef, no-shadow
  } else {
    attr.setWithWorker({
      current: parseFloat(val),
    }); // .set()
  }
};

const changeMoney = (startamount, addamount) => {
  //! changeMoney
  if (addamount !== null) {
    let total = startamount;

    const currency = addamount.slice(-2);
    const amount2 = -parseFloat(addamount.substr(0, addamount.length - 2));
    const origamount = total;
    let amount3 = 0;
    if (currency === 'cp') {
      amount3 = amount2 / 100;
    }
    if (currency === 'sp') {
      amount3 = amount2 / 10;
    }
    if (currency === 'ep') {
      amount3 = amount2 / 2;
    }
    if (currency === 'gp') {
      amount3 = amount2;
    }
    if (currency === 'pp') {
      amount3 = amount2 * 10;
    }
    if (
      (total[0] * 10) +
      total[1] +
      (total[2] / 2) +
      (total[3] / 10) +
      (total[4] / 100)
      >= -amount3
    ) {
      total[4] += amount3 * 100;
      while (total[4] < 0) {
        total[4] += 10;
        total[3] -= 1;
      } // cp
      while (total[3] < 0) {
        if (total[4] >= 10) {
          total[4] -= 10;
          total[3] += 1;
        } else {
          total[3] += 5;
          total[2] -= 1;
        }
      } // sp
      while (total[2] < 0) {
        if (total[3] >= 5) {
          total[3] -= 5;
          total[2] += 1;
        } else {
          total[2] += 2;
          total[1] -= 1;
        }
      } // ep
      while (total[1] < 0) {
        if (total[2] >= 2) {
          total[2] -= 2;
          total[1] += 1;
        } else {
          total[1] += 10;
          total[0] -= 1;
        }
      } // gp
      while (total[0] < 0) {
        if (total[1] >= 10) {
          total[1] -= 10;
          total[0] += 1;
        } else {
          total = origamount;
          return 'ERROR: Not enough cash.';
        }
      } // pp
      return total;
    }
    return 'ERROR: Not enough cash.';
  }
  return 0;
};

const toUsd = (total, usd = 110) => {
  //! toUsd
  let output = '';
  if (usd > 0) {
    output = `${total} gp <small>(~ ${(Math.round((total * usd) / 5) * 5)} USD)</small>`;
  } else {
    output = `${total} gp`;
  }
  return output;
};

const myoutput = (character, usd = 110) => {
  //! myoutput

  const name = getAttrByName(character.id, 'character_name');
  const pp = parseFloat(getattr(character.id, 'pp')) || 0;
  const gp = parseFloat(getattr(character.id, 'gp')) || 0;
  const ep = parseFloat(getattr(character.id, 'ep')) || 0;
  const sp = parseFloat(getattr(character.id, 'sp')) || 0;
  const cp = parseFloat(getattr(character.id, 'cp')) || 0;
  const total = Math.round((
    (pp * 10) +
  (ep * 0.5) +
  gp +
  (sp / 10) +
  (cp / 100)
  ) * 10000) / 10000;
  const weight = (pp + gp + ep + sp + cp) / 50;

  let output = `${name}: <b>${toUsd(total, usd)}</b><br><small>`;
  if (pp) output += `<em style='color:blue;'>${pp} pp</em>, `;
  if (gp) output += `<em style='color:orange;'>${gp} gp</em>, `;
  if (ep) output += `<em style='color:silver;'>${ep} ep</em>, `;
  if (sp) output += `<em style='color:grey;'>${sp} sp</em>, `;
  if (cp) output += `<em style='color:brown;'>${cp} cp</em>`;

  output += `<br>(${weight} lbs)</small><br><br>`;
  return [output, total];
};


on('ready', () => {
  const v = '%%version%%'; // version number
  const usd = 110;
  /*
  Change this if you want to have a rough estimation of a character’s wealth in USD.
  After some research I believe a reasonable exchange ratio is roughly 1 gp = 110 USD
  Set it to 0 to disable it completely.
  */

  const scname = 'CashMaster'; // script name
  let selectedsheet = 'OGL'; // You can set this to "5E-Shaped" if you're using the Shaped sheet

  // detecting useroptions from one-click
  if (globalconfig && globalconfig.cashmaster && globalconfig.cashmaster.useroptions) {
    selectedsheet = globalconfig.cashmaster.useroptions.selectedsheet; // eslint-disable-line prefer-destructuring
  }
  let rt = '';
  if (selectedsheet === 'OGL') {
    rt = ['desc', 'desc'];
  } else if (selectedsheet === '5E-Shaped') {
    rt = ['5e-shaped', 'text'];
  } else {
    rt = ['default', `name=${scname} }}{{note`];
  }


  log(`${scname} v${v} online. Select one or more party members, then use \`!cm -help\``);

  let pp;
  let gp;
  let ep;
  let sp;
  let cp;
  let total;
  let output;
  let ppa;
  let gpa;
  let epa;
  let spa;
  let cpa;
  let ppg;
  let gpg;
  let epg;
  let spg;
  let cpg;
  let name;
  let usd2;

  on('chat:message', (msg) => {
    if (msg.type !== 'api' && !playerIsGM(msg.playerid)) return;
    if (msg.content.startsWith('!cm') !== true) return;
    if (msg.selected == null) {
      sendChat(scname, '/w gm **ERROR:** You need to select at least one character.');
      return;
    }

    let partytotal = 0;
    let partycounter = 0;
    const partymember = Object.entries(msg.selected).length;
    msg.selected.forEach((obj) => {
      const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
      let character;
      if (token) {
        character = getObj('character', token.get('represents'));
      }
      if (character) {
        partycounter += 1;
        name = getAttrByName(character.id, 'character_name');
        pp = parseFloat(getattr(character.id, 'pp')) || 0;
        gp = parseFloat(getattr(character.id, 'gp')) || 0;
        ep = parseFloat(getattr(character.id, 'ep')) || 0;
        sp = parseFloat(getattr(character.id, 'sp')) || 0;
        cp = parseFloat(getattr(character.id, 'cp')) || 0;
        total = Math.round((
          (pp * 10) +
          (ep * 0.5) +
          gp +
          (sp / 10) +
          (cp / 100)
        ) * 10000) / 10000;
        partytotal = total + partytotal;
      }
    });

    partytotal = Math.round(partytotal * 100, 0) / 100;

    if (msg.content.includes('-help') || msg.content === '!cm' || msg.content.includes('-h')) {
      //! help
      sendChat(scname, `/w gm %%README%%`); // eslint-disable-line quotes
    }

    if (msg.content.includes('-share') || msg.content.includes('-best-share') || msg.content.includes('-s') || msg.content.includes('-bs')) {
      //! share and convert
      output = '';
      const cashshare = partytotal / partycounter;
      let newcounter = 0;
      let pps = Math.floor(cashshare / 10);
      if (msg.content.includes('-share') || msg.content.includes('-s')) {
        pps = 0;
      }
      let rest = cashshare - (pps * 10);
      const gps = Math.floor(rest);
      rest = (rest - gps) * 2;
      let eps = Math.floor(rest);
      if (msg.content.includes('-share') || msg.content.includes('-s')) {
        eps = 0;
      }
      rest = (rest - eps) * 5;
      const sps = Math.floor(rest);
      rest = (rest - sps) * 10;
      let cps = Math.round(rest);
      rest = (rest - cps) * partycounter;

      sendChat(scname, `/w gm &{template:${rt[0]}} {{${rt[1]}=<b>Let’s share this!</b><hr>Everyone receives the equivalent of ${toUsd(cashshare)} gp: ${pps} platinum, ${gps} gold, ${eps} electrum, ${sps} silver, and ${cps} copper.}}`);

      msg.selected.forEach((obj) => {
        const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
        let character;
        newcounter += 1;
        if (token) {
          character = getObj('character', token.get('represents'));
        }
        if (character) {
          setattr(character.id, 'pp', pps);
          setattr(character.id, 'gp', gps);
          setattr(character.id, 'ep', eps);
          setattr(character.id, 'sp', sps);
          // enough copper coins? If not, the last one in the group has to take the diff
          if ((rest > 0.999 || rest < -0.999) && newcounter === partycounter) {
            cps += Math.round(rest);
          }
          setattr(character.id, 'cp', cps);
        }
      });
    }

    if (msg.content.includes('-add') || msg.content.includes('-a')) {
      //! add
      ppg = /([0-9 -]+)pp/;
      ppa = ppg.exec(msg.content);

      gpg = /([0-9 -]+)gp/;
      gpa = gpg.exec(msg.content);

      epg = /([0-9 -]+)ep/;
      epa = epg.exec(msg.content);

      spg = /([0-9 -]+)sp/;
      spa = spg.exec(msg.content);

      cpg = /([0-9 -]+)cp/;
      cpa = cpg.exec(msg.content);

      output = '';

      msg.selected.forEach((obj) => {
        const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
        let character;
        if (token) {
          character = getObj('character', token.get('represents'));
        }
        if (character) {
          partycounter += 1;
          name = getAttrByName(character.id, 'character_name');
          pp = parseFloat(getattr(character.id, 'pp')) || 0;
          gp = parseFloat(getattr(character.id, 'gp')) || 0;
          ep = parseFloat(getattr(character.id, 'ep')) || 0;
          sp = parseFloat(getattr(character.id, 'sp')) || 0;
          cp = parseFloat(getattr(character.id, 'cp')) || 0;
          total = Math.round((
            (pp * 10) +
            (ep * 0.5) +
            gp +
            (sp / 10) +
            (cp / 100)
          ) * 10000) / 10000;
          partytotal = total + partytotal;
          output += `<br><b>${name}</b>`;
          if (ppa) {
            setattr(character.id, 'pp', parseFloat(pp) + parseFloat(ppa[1]));
            output += `<br> ${ppa[0]}`;
          }
          if (gpa) {
            setattr(character.id, 'gp', parseFloat(gp) + parseFloat(gpa[1]));
            output += `<br> ${gpa[0]}`;
          }
          if (epa) {
            setattr(character.id, 'ep', parseFloat(ep) + parseFloat(epa[1]));
            output += `<br> ${epa[0]}`;
          }
          if (spa) {
            setattr(character.id, 'sp', parseFloat(sp) + parseFloat(spa[1]));
            output += `<br> ${spa[0]}`;
          }
          if (cpa) {
            setattr(character.id, 'cp', parseFloat(cp) + parseFloat(cpa[1]));
            output += `<br> ${cpa[0]}`;
          }
        }
      });
      sendChat(scname, `/w gm &{template:${rt[0]}} {{${rt[1]}=<b>Cashing out - it’s payday!</b><hr>${output}}}`);
    }

    if (msg.content.includes('-pay') || msg.content.includes('-p')) {
      //! pay
      ppg = /([0-9 -]+)pp/;
      ppa = ppg.exec(msg.content);

      gpg = /([0-9 -]+)gp/;
      gpa = gpg.exec(msg.content);

      epg = /([0-9 -]+)ep/;
      epa = epg.exec(msg.content);

      spg = /([0-9 -]+)sp/;
      spa = spg.exec(msg.content);

      cpg = /([0-9 -]+)cp/;
      cpa = cpg.exec(msg.content);

      output = '';

      msg.selected.forEach((obj) => {
        const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
        let character;
        if (token) {
          character = getObj('character', token.get('represents'));
        }
        if (character) {
          partycounter += 1;
          name = getAttrByName(character.id, 'character_name');
          pp = parseFloat(getattr(character.id, 'pp')) || 0;
          gp = parseFloat(getattr(character.id, 'gp')) || 0;
          ep = parseFloat(getattr(character.id, 'ep')) || 0;
          sp = parseFloat(getattr(character.id, 'sp')) || 0;
          cp = parseFloat(getattr(character.id, 'cp')) || 0;

          let startamount = [pp, gp, ep, sp, cp];
          if (ppa !== null) startamount = changeMoney(startamount, ppa[0]);
          if (gpa !== null) startamount = changeMoney(startamount, gpa[0]);
          if (epa !== null) startamount = changeMoney(startamount, epa[0]);
          if (spa !== null) startamount = changeMoney(startamount, spa[0]);
          if (cpa !== null) startamount = changeMoney(startamount, cpa[0]);

          output += `<br><b>${name}</b> has `;
          if (startamount === 'ERROR: Not enough cash.') output += 'not enough cash!';
          else {
            setattr(character.id, 'pp', parseFloat(startamount[0]));
            output += `<br> ${startamount[0]}pp`;
            setattr(character.id, 'gp', parseFloat(startamount[1]));
            output += `<br> ${startamount[1]}gp`;
            setattr(character.id, 'ep', parseFloat(startamount[2]));
            output += `<br> ${startamount[2]}ep`;
            setattr(character.id, 'sp', parseFloat(startamount[3]));
            output += `<br> ${startamount[3]}sp`;
            setattr(character.id, 'cp', parseFloat(startamount[4]));
            output += `<br> ${startamount[4]}cp`;
          }
        }
      });
      sendChat(scname, `/w gm &{template:${rt[0]}} {{${rt[1]}=<b>Cashing out - it’s payday!</b><hr>${output}}}`);
    }

    if (msg.content.includes('-loot') || msg.content.includes('-l')) {
      //! loot
      ppg = /([0-9 -]+)pp/;
      ppa = ppg.exec(msg.content);

      gpg = /([0-9 -]+)gp/;
      gpa = gpg.exec(msg.content);

      epg = /([0-9 -]+)ep/;
      epa = epg.exec(msg.content);

      spg = /([0-9 -]+)sp/;
      spa = spg.exec(msg.content);

      cpg = /([0-9 -]+)cp/;
      cpa = cpg.exec(msg.content);

      output = '';
      partycounter = 0;

      msg.selected.forEach((obj) => {
        const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
        let character;
        if (token) {
          character = getObj('character', token.get('represents'));
        }
        if (character) {
          partycounter += 1;
          name = getAttrByName(character.id, 'character_name');
          pp = parseFloat(getattr(character.id, 'pp')) || 0;
          gp = parseFloat(getattr(character.id, 'gp')) || 0;
          ep = parseFloat(getattr(character.id, 'ep')) || 0;
          sp = parseFloat(getattr(character.id, 'sp')) || 0;
          cp = parseFloat(getattr(character.id, 'cp')) || 0;

          let ppt;
          let gpt;
          let ept;
          let spt;
          let cpt;

          if (ppa !== null) {
            ppt = cashsplit(ppa[1], partymember, partycounter);
          }
          if (gpa !== null) {
            gpt = cashsplit(gpa[1], partymember, partycounter);
          }
          if (epa !== null) {
            ept = cashsplit(epa[1], partymember, partycounter);
          }
          if (spa !== null) {
            spt = cashsplit(spa[1], partymember, partycounter);
          }
          if (cpa !== null) {
            cpt = cashsplit(cpa[1], partymember, partycounter);
          }

          output += `<br><b>${name}</b>`;
          if (ppa) {
            setattr(character.id, 'pp', parseFloat(pp) + parseFloat(ppt));
            output += `<br> ${ppt}pp`;
          }
          if (gpa) {
            setattr(character.id, 'gp', parseFloat(gp) + parseFloat(gpt));
            output += `<br> ${gpt}gp`;
          }
          if (epa) {
            setattr(character.id, 'ep', parseFloat(ep) + parseFloat(ept));
            output += `<br> ${ept}ep`;
          }
          if (spa) {
            setattr(character.id, 'sp', parseFloat(sp) + parseFloat(spt));
            output += `<br> ${spt}sp`;
          }
          if (cpa) {
            setattr(character.id, 'cp', parseFloat(cp) + parseFloat(cpt));
            output += `<br> ${cpt}cp`;
          }
        }
      });
      sendChat(scname, `/w gm &{template:${rt[0]}} {{${rt[1]}=<b>You are splitting up the coins among you</b><hr>${output}}}`);
    }

    if (msg.content.includes('-add') || msg.content.includes('-pay') || msg.content.includes('-share') || msg.content.includes('-best-share') || msg.content.includes('-loot') || msg.content.includes('-overview') || msg.content.includes('-a') || msg.content.includes('-p') || msg.content.includes('-s') || msg.content.includes('-bs') || msg.content.includes('-l') || msg.content.includes('-o')) {
      //! overview
      partytotal = 0;
      partycounter = 0;
      if (!msg.content.includes('--usd')) usd2 = 0;
      else usd2 = usd;
      output = `/w gm &{template:${rt[0]}} {{${rt[1]}=<b>Party’s cash overview</b><br><br>`;
      msg.selected.forEach((obj) => {
        const token = getObj('graphic', obj._id); // eslint-disable-line no-underscore-dangle
        let character;
        if (token) {
          character = getObj('character', token.get('represents'));
        }
        if (character) {
          output += myoutput(character, usd2)[0];
          partytotal += myoutput(character, usd2)[1];
        }
      });
      partytotal = Math.round(partytotal * 100, 0) / 100;

      output += `<b><u>Party total: ${toUsd(partytotal, usd2)}</u></b>}}`;
      sendChat(scname, output);
    }
  });
});
